include/LinearAlgebra/sparseLUSolver.h

```diff
@@ -103,206 +99,465 @@ SparseLUSolver<T>::SparseLUSolver(T tolerance_abs, T tolerance_rel)
+void SparseLUSolver<T>::symbolicFactorization(const SparseMatrixCSR<T>& A, std::vector<std::vector<int>>& L_pattern,
+                                              std::vector<std::vector<int>>& U_pattern) const
 {
-    assert(factorized_);
-    const int n = L_row_ptr.size() - 1; // n is the number of rows in the matrix
+    const int n = A.rows();
+    L_pattern.clear();
+    U_pattern.clear();
+    L_pattern.resize(n);
+    U_pattern.resize(n);
+
+    // Marker array tracks visited columns per row
+    std::vector<int> marker(n, -1);
+    std::vector<int> row_marked_indices;
+    std::vector<int> stk;
+
+    for (int i = 0; i < n; ++i) {
+        row_marked_indices.clear();
+        stk.clear();
+
+        // Process original non-zeros in row i
+        const int nnz = A.row_nz_size(i);
+        for (int idx = 0; idx < nnz; ++idx) {
+            int col = A.row_nz_index(i, idx);
+            if (marker[col] != i) {
+                marker[col] = i;
+                row_marked_indices.push_back(col);
+                if (col < i) { // Lower triangular element
+                    stk.push_back(col);
+                }
+            }
+        }
 
-    // Forward substitution (L * b = b) -> b now holds y
-    for (int i = 0; i < n; i++) {
-        for (int idx = L_row_ptr[i]; idx < L_row_ptr[i + 1]; idx++) {
-            b[i] -= L_values[idx] * b[L_col_idx[idx]];
+        // Ensure diagonal is included
+        if (marker[i] != i) {
+            marker[i] = i;
+            row_marked_indices.push_back(i);
         }
-    }
 
-    // Backward substitution (U * b = y) -> b now holds x
-    for (int i = n - 1; i >= 0; i--) {
-        T diag = 0;
-        for (int idx = U_row_ptr[i]; idx < U_row_ptr[i + 1]; idx++) {
-            int col = U_col_idx[idx];
-            if (col == i) {
-                diag = U_values[idx]; // Store diagonal value
+        // Process fill-in elements
+        while (!stk.empty()) {
+            int j = stk.back();
+            stk.pop_back();
+
+            // Process U[j] pattern
+            for (int col2 : U_pattern[j]) {
+                if (marker[col2] != i) {
+                    marker[col2] = i;
+                    row_marked_indices.push_back(col2);
+                    if (col2 < i) { // Continue for lower indices
+                        stk.push_back(col2);
+                    }
+                }
+            }
+        }
+
+        // Split into L and U patterns
+        auto& Li = L_pattern[i];
+        auto& Ui = U_pattern[i];
+        for (int col : row_marked_indices) {
+            if (col < i) {
+                Li.push_back(col);
             }
             else {
-                b[i] -= U_values[idx] * b[col];
+                Ui.push_back(col);
             }
         }
-        assert(std::abs(diag) > tolerance_abs_);
-        b[i] /= diag;
+
+        // Sort patterns for efficient access
+        std::sort(Li.begin(), Li.end());
+        std::sort(Ui.begin(), Ui.end());
```

Comment:
I don't understand what this step is for. Is the order really arbitrary? Or does each array contain a diagonal and it is sorting by length thereby leaving the array untouched?

Reply:
Ah ok. I read this as sorting L not Li. I need to think more about the algorithm to work out why they're out of order in the first place though.

Reply:
After investigation using this test:
https://github.com/SciCompMod/GMGPolar/blob/7e23dbb6a075ff931de903e59fab4f5656fb0e34/tests/LinearAlgebra/csr_solver.cpp#L146-L163

It looks like the sorting is needed because the results of `A.row_nz_index(i, idx)` are not necessarily sorted